platforms/macos/src/subclass.rs

```diff
@@ -0,0 +1,183 @@
+        // the object is created.
+        adapter: IvarDrop<Box<Adapter>>,
+        prev_class: &'static Class,
+    }
+
+    unsafe impl ClassType for AssociatedObject {
+        type Super = NSObject;
+        const NAME: &'static str = "AccessKitSubclassAssociatedObject";
+    }
+);
+
+impl AssociatedObject {
+    fn new(adapter: Adapter, prev_class: &'static Class) -> Id<Self, Owned> {
+        unsafe {
+            let mut object: Id<Self, Owned> = msg_send_id![Self::class(), new];
+            Ivar::write(&mut object.adapter, Box::new(adapter));
+            Ivar::write(&mut object.prev_class, prev_class);
+            object
+        }
+    }
+}
+
+fn associated_object(view: &NSView) -> &AssociatedObject {
+    unsafe {
+        (objc_getAssociatedObject(view as *const NSView as *const _, associated_object_key())
+            as *const AssociatedObject)
+            .as_ref()
+    }
+    .unwrap()
+}
+
+// Some view classes, like the one in winit 0.27, assume that they are the
+// lowest subclass, and call [self superclass] to get their superclass.
+// Give them the answer they need.
+unsafe extern "C" fn superclass(this: &NSView, _cmd: Sel) -> Option<&Class> {
+    let associated = associated_object(this);
+    associated.prev_class.superclass()
+}
+
+unsafe extern "C" fn children(this: &NSView, _cmd: Sel) -> *mut NSArray<NSObject> {
+    let associated = associated_object(this);
+    associated.adapter.view_children()
+}
+
+unsafe extern "C" fn focus(this: &NSView, _cmd: Sel) -> *mut NSObject {
+    let associated = associated_object(this);
+    associated.adapter.focus()
+}
+
+unsafe extern "C" fn hit_test(this: &NSView, _cmd: Sel, point: NSPoint) -> *mut NSObject {
+    let associated = associated_object(this);
+    associated.adapter.hit_test(point)
+}
+
+/// Uses dynamic Objective-C subclassing to implement the NSView
+/// accessibility methods when normal subclassing isn't an option.
+pub struct SubclassingAdapter {
+    view: WeakId<NSView>,
+    associated: Id<AssociatedObject, Owned>,
+}
+
+impl SubclassingAdapter {
+    /// Dynamically subclass the specified view to use the specified adapter.
+    ///
+    /// # Safety
+    ///
+    /// `view` must be a valid, unreleased pointer to an `NSView`.
+    pub unsafe fn new(view: *mut c_void, adapter: Adapter) -> Self {
+        let view = view as *mut NSView;
+        // Cast to a pointer and back to force the lifetime to 'static
+        // SAFETY: We know the class will live as long as the instance,
+        // and we only use this reference while the instance is alive.
+        let prev_class = unsafe { &*((*view).class() as *const Class) };
+        let mut associated = AssociatedObject::new(adapter, prev_class);
+        unsafe {
+            objc_setAssociatedObject(
+                view as *mut _,
+                associated_object_key(),
+                Id::as_mut_ptr(&mut associated) as *mut _,
+                OBJC_ASSOCIATION_ASSIGN,
+            )
```

Comment:
I know that `objc_setAssociatedObject` and `objc_getAssociatedObject` takes and returns an `id`/`objc_object*`, but if you're using `OBJC_ASSOCIATION_ASSIGN`, then it can actually be a pointer to anything - so you didn't need to make `AssociatedObject` a class, you could simply have put it in a `Box`.

Reply:
However, now that you have done so, it can actually be to our benefit:

The code here is a bit dangerous, since if `SubclassingAdapter.associated` (which holds the last retain count to the object) is dropped before the associated object is removed from the view, the view will access a released pointer - of course, this can't happen currently, the implementation of `Drop` ensures that, but still something to keep in mind!

You could consider using `OBJC_ASSOCIATION_RETAIN` instead, which will make the object be retained by the view, which will ensure that it is released only once the view is.

Just a consideration, using `Box` and documenting that the code in `Drop` is necessary for soundness is also fine.